File: clients/unity/DedicatedProtocolSmokeTest.cs
// Unity sample: dedicated-server protocol smoke test.
//
// Drop this file into a Unity project (Assets/Tests/, assembly with
// "UnityEngine.TestRunner" + "System.Net.WebSockets" available; Unity 2021+)
// and run it from the Test Runner while `npm run dedicated` is running, or
// point DEDICATED_URL elsewhere. No game rendering is attempted here — this
// verifies the same wire contract that scripts/dedicated-check.mjs enforces,
// so a future Unity viewer starts from a known-good connection layer.
//
// Honest status: written against dedicated/PROTOCOL.md and kept in sync with
// the Node conformance check; it compiles only inside a Unity project and was
// not executed on the machine that authored it (no Unity installed there).

using System;
using System.Collections;
using System.Net.WebSockets;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using NUnit.Framework;
using UnityEngine.TestTools;

public class DedicatedProtocolSmokeTest
{
    private const string DefaultUrl = "ws://127.0.0.1:8643/";

    private static string ServerUrl =>
        Environment.GetEnvironmentVariable("DEDICATED_URL") ?? DefaultUrl;

    [UnityTest]
    public IEnumerator ViewerReceivesWelcomeAndSnapshot()
    {
        var task = RunViewerFlow();
        while (!task.IsCompleted) yield return null;
        if (task.IsFaulted) throw task.Exception.GetBaseException();
    }

    private static async Task RunViewerFlow()
    {
        using var socket = new ClientWebSocket();
        using var cancel = new CancellationTokenSource(TimeSpan.FromSeconds(15));
        await socket.ConnectAsync(new Uri(ServerUrl), cancel.Token);

        await Send(socket, "{\"type\":\"hello\",\"role\":\"viewer\"}", cancel.Token);

        // 1) welcome: protocol v1, role viewer.
        string welcome = await Receive(socket, cancel.Token);
        StringAssert.Contains("\"type\":\"welcome\"", welcome);
        StringAssert.Contains("\"protocol\":1", welcome);
        StringAssert.Contains("\"role\":\"viewer\"", welcome);

        // 2) first snapshot: documented schema fields exist.
        string snapshot = await ReceiveOfType(socket, "\"type\":\"snapshot\"", cancel.Token);
        foreach (string field in new[] { "\"tick\"", "\"phase\"", "\"wave\"", "\"castleHp\"", "\"enemies\"", "\"board\"" })
            StringAssert.Contains(field, snapshot);

        // 3) commands are refused without the controller key.
        await Send(socket, "{\"type\":\"command\",\"op\":\"pause\"}", cancel.Token);
        string refusal = await ReceiveOfType(socket, "\"code\":\"forbidden\"", cancel.Token);
        StringAssert.Contains("\"type\":\"error\"", refusal);

        await socket.CloseAsync(WebSocketCloseStatus.NormalClosure, "done", cancel.Token);
    }

    private static Task Send(ClientWebSocket socket, string json, CancellationToken token) =>
        socket.SendAsync(new ArraySegment<byte>(Encoding.UTF8.GetBytes(json)),
            WebSocketMessageType.Text, true, token);

    private static async Task<string> Receive(ClientWebSocket socket, CancellationToken token)
    {
        var buffer = new byte[512 * 1024];
        var builder = new StringBuilder();
        while (true)
        {
            var result = await socket.ReceiveAsync(new ArraySegment<byte>(buffer), token);
            builder.Append(Encoding.UTF8.GetString(buffer, 0, result.Count));
            if (result.EndOfMessage) return builder.ToString();
        }
    }

    private static async Task<string> ReceiveOfType(ClientWebSocket socket, string marker, CancellationToken token)
    {
        // Skip unrelated broadcasts (events/decisions) until the marker arrives.
        for (int i = 0; i < 200; i++)
        {
            string message = await Receive(socket, token);
            if (message.Contains(marker)) return message;
        }
        throw new TimeoutException($"No message containing {marker}");
    }
}

File: clients/unreal/DedicatedProtocolSmokeTest.cpp
// Unreal sample: dedicated-server protocol smoke test.
//
// Drop into an Unreal 5 project's test source (e.g. Source/<Game>/Tests/) with
// "WebSockets" and "Json" in the module's Build.cs dependencies, then run from
// Session Frontend → Automation while `npm run dedicated` is running.
// This verifies the same wire contract as scripts/dedicated-check.mjs so a
// future Unreal viewer starts from a known-good connection layer.
//
// Honest status: written against dedicated/PROTOCOL.md and kept in sync with
// the Node conformance check; it compiles only inside an Unreal project and
// was not executed on the machine that authored it (no Unreal installed there).

#include "CoreMinimal.h"
#include "Dom/JsonObject.h"
#include "IWebSocket.h"
#include "Misc/AutomationTest.h"
#include "Serialization/JsonReader.h"
#include "Serialization/JsonSerializer.h"
#include "WebSocketsModule.h"

IMPLEMENT_SIMPLE_AUTOMATION_TEST(FDedicatedProtocolSmokeTest,
    "ConstellationDefense.Dedicated.ProtocolSmoke",
    EAutomationTestFlags::EditorContext | EAutomationTestFlags::ProductFilter)

namespace
{
    TSharedPtr<FJsonObject> ParseJson(const FString& Text)
    {
        TSharedPtr<FJsonObject> Object;
        const TSharedRef<TJsonReader<>> Reader = TJsonReaderFactory<>::Create(Text);
        FJsonSerializer::Deserialize(Reader, Object);
        return Object;
    }
}

bool FDedicatedProtocolSmokeTest::RunTest(const FString& Parameters)
{
    const FString Url = FPlatformMisc::GetEnvironmentVariable(TEXT("DEDICATED_URL")).IsEmpty()
        ? TEXT("ws://127.0.0.1:8643/")
        : FPlatformMisc::GetEnvironmentVariable(TEXT("DEDICATED_URL"));

    TSharedPtr<IWebSocket> Socket = FWebSocketsModule::Get().CreateWebSocket(Url);

    bool bWelcomeOk = false;      // welcome with protocol 1 and role viewer
    bool bSnapshotOk = false;     // snapshot carrying the documented fields
    bool bForbiddenOk = false;    // command refused without the controller key
    bool bDone = false;

    Socket->OnConnected().AddLambda([&]
    {
        Socket->Send(TEXT("{\"type\":\"hello\",\"role\":\"viewer\"}"));
    });

    Socket->OnMessage().AddLambda([&](const FString& Message)
    {
        const TSharedPtr<FJsonObject> Json = ParseJson(Message);
        if (!Json.IsValid()) { return; }
        const FString Type = Json->GetStringField(TEXT("type"));

        if (Type == TEXT("welcome"))
        {
            bWelcomeOk = Json->GetIntegerField(TEXT("protocol")) == 1
                && Json->GetStringField(TEXT("role")) == TEXT("viewer");
        }
        else if (Type == TEXT("snapshot") && !bSnapshotOk)
        {
            bSnapshotOk = Json->HasField(TEXT("tick")) && Json->HasField(TEXT("phase"))
                && Json->HasField(TEXT("wave")) && Json->HasField(TEXT("castleHp"))
                && Json->HasField(TEXT("enemies")) && Json->HasField(TEXT("board"));
            // Snapshot in hand: prove the auth boundary next.
            Socket->Send(TEXT("{\"type\":\"command\",\"op\":\"pause\"}"));
        }
        else if (Type == TEXT("error"))
        {
            if (Json->GetStringField(TEXT("code")) == TEXT("forbidden"))
            {
                bForbiddenOk = true;
                bDone = true;
            }
        }
    });

    Socket->OnConnectionError().AddLambda([&](const FString& Error)
    {
        AddError(FString::Printf(TEXT("Connection failed: %s (is `npm run dedicated` running?)"), *Error));
        bDone = true;
    });

    Socket->Connect();

    // Automation tests may block briefly; 15 s covers a cold local server.
    const double Deadline = FPlatformTime::Seconds() + 15.0;
    while (!bDone && FPlatformTime::Seconds() < Deadline)
    {
        FPlatformProcess::Sleep(0.05f);
    }
    Socket->Close();

    TestTrue(TEXT("welcome carried protocol v1 and viewer role"), bWelcomeOk);
    TestTrue(TEXT("snapshot carried the documented schema"), bSnapshotOk);
    TestTrue(TEXT("viewer command was refused (forbidden)"), bForbiddenOk);
    return bWelcomeOk && bSnapshotOk && bForbiddenOk;
}
